### PR TITLE
[DOCS] Adds info about support for G1GC. Closes #43885

### LINKED ISSUE #43885: Clarify G1GC support on docs
As of 6.5.0, [G1GC is now supported by Elasticsearch](https://www.elastic.co/blog/elastic-stack-6-5-0-released).

We should document this, perhaps in the [setup page](https://www.elastic.co/guide/en/elasticsearch/reference/current/setup.html).

### PR DESCRIPTION
**Pull request by debadair**
Applies to v6.5+.

### CODE REVIEW

File: docs/reference/setup.asciidoc [hunk 1]
```diff
@@ -35,6 +35,27 @@ http://www.oracle.com/technetwork/java/eol-135779.html[LTS version of Java].
...
+== JVM garbage collectors
+
+The default garbage collector for {es} is Concurrent-Mark and Sweep (CMS). 
+This GC runs concurrently with the execution of the application so that it
+can minimize pauses. It does, however, have two stop-the-world phases when 
+the JVM literally halts execution so it can trace the object graph and
+collect dead objects. CMS also has trouble collecting large heaps.
+
+{es} also supports the newer Garbage First GC (G1GC). G1GC minimizes pauses 
+and operates better on large heaps than CMS. To use G1GC, you must run JDK 8u40
```
Comment (dliappis) [inline docs/reference/setup.asciidoc]: > operates better on large heaps than CMS

This is a bit open scoped and IMHO I don't think it's very useful without more data (see comment below too).

File: docs/reference/setup.asciidoc [hunk 2]
```diff
@@ -35,6 +35,27 @@ http://www.oracle.com/technetwork/java/eol-135779.html[LTS version of Java].
...
 
+[float]
+[[garbage-collectors]]
+== JVM garbage collectors
+
+The default garbage collector for {es} is Concurrent-Mark and Sweep (CMS). 
+This GC runs concurrently with the execution of the application so that it
+can minimize pauses. It does, however, have two stop-the-world phases when 
+the JVM literally halts execution so it can trace the object graph and
+collect dead objects. CMS also has trouble collecting large heaps.
```
Comment (dliappis) [inline docs/reference/setup.asciidoc]: > CMS also has trouble collecting large heaps.

This is a big generic and might lead users to just switch to G1GC without a second thought in most cases. Might be better explained in a more detailed article elsewhere and not mentioned here?

**Review by dliappis — COMMENTED** [review]
Thanks! I am a bit skeptical about making recommendations esp. regarding JVM internals in our docs and left some comments.

@dakrone thoughts?

**Comment by debadair** [comment]
Mentioning that G1GC is supported invites the question of when it should be used. (And questions about G1GC come up pretty frequently, anyway.) IMO, the docs *are* the right place to address that question--but I agree that it's a more involved topic. I think the right thing to do here is prune out the "commentary" on the garbage collectors and make it clear that the default is still the recommended choice. We can tackle the more nuanced discussion separately.

**Comment by dliappis** [comment]
> I think the right thing to do here is prune out the "commentary" on the garbage collectors and make it clear that the default is still the recommended choice. We can tackle the more nuanced discussion separately.

This sounds like a good approach.
Another thing is minimum supportable Java for G1GC. Based [on this discussion](https://discuss.elastic.co/t/where-can-i-find-info-on-the-status-of-g1gc-in-lucene-and-es/149526/7?u=dliappis) I think we are safer with Java 11 onwards.

**Comment by dliappis** [comment]
> I think the right thing to do here is prune out the "commentary" on the garbage collectors and make it clear that the default is still the recommended choice. We can tackle the more nuanced discussion separately.

@debadair not intended to pressure, just wanted to clarify if there are additional clarifications required from my side for rewording this based on the above sentence.

**Comment by dliappis** [comment]
@debadair would you have time to take a look at this based on [the latest feedback](https://github.com/elastic/elasticsearch/pull/44145#issuecomment-510859223), or is there something I could do on my side to help move this forward?

**Comment by debadair** [comment]
Closing this PR, but we need to provide updated information about garbage collection.

**Pull request closed without merging**